go/chat/storage/inbox.go

```diff
@@ -172,130 +193,150 @@ func (i *Inbox) readDiskInbox(ctx context.Context, uid gregor1.UID, useInMemory
-		ibox.InboxVersion, ibox.Version, ibox.ServerVersion, len(ibox.Conversations))
-	if ierr := i.writeDiskBox(ctx, i.dbKey(uid), ibox); ierr != nil {
-		return NewInternalError(ctx, i.DebugLabeler, "failed to write inbox: uid: %s err: %s", uid, ierr)
+func (i *Inbox) writeDiskIndex(ctx context.Context, uid gregor1.UID, ibox inboxDiskIndex) Error {
+	i.Debug(ctx, "writeDiskIndex: convs: %d queries: %d", len(ibox.ConversationIDs), len(ibox.Queries))
+	inboxMemCache.PutIndex(uid, &ibox)
+	if err := i.writeDiskBox(ctx, i.dbIndexKey(uid), ibox); err != nil {
+		return NewInternalError(ctx, i.DebugLabeler, "failed to write inbox index: %s", err)
 	}
-	i.writeMobileSharedInbox(ctx, *ibox, uid)
 	return nil
 }
 
-func (i *Inbox) Flush(ctx context.Context, uid gregor1.UID) (err Error) {
-	locks.Inbox.Lock()
-	defer locks.Inbox.Unlock()
-	defer i.Trace(ctx, func() error { return err }, fmt.Sprintf("Flush(%s)", uid))()
-	return i.flushLocked(ctx, uid)
+func (i *Inbox) readConvs(ctx context.Context, uid gregor1.UID, convIDs []chat1.ConversationID) (res []types.RemoteConversation, err Error) {
+	res = make([]types.RemoteConversation, 0, len(convIDs))
+	memHits := make(map[chat1.ConvIDStr]bool, len(convIDs))
+	for _, convID := range convIDs {
+		if conv := inboxMemCache.GetConv(uid, convID); conv != nil {
+			res = append(res, *conv)
+			memHits[convID.ConvIDStr()] = true
+		}
+	}
+	for _, convID := range convIDs {
+		if memHits[convID.ConvIDStr()] {
+			continue
+		}
+		var conv types.RemoteConversation
+		found, err := i.readDiskBox(ctx, i.dbConvKey(uid, convID), &conv)
+		if err != nil {
+			if _, ok := err.(libkb.LoginRequiredError); ok {
+				return res, MiscError{Msg: err.Error()}
+			}
+			return res, NewInternalError(ctx, i.DebugLabeler,
+				"failed to read inbox: uid: %d err: %s", uid, err)
+		}
+		if !found {
+			return res, MissError{}
+		}
+		res = append(res, conv)
+	}
+	return res, nil
 }
 
-func (i *Inbox) writeDiskInbox(ctx context.Context, uid gregor1.UID, ibox inboxDiskData) Error {
-	// Get latest server version
-	vers, err := i.G().ServerCacheVersions.Fetch(ctx)
+func (i *Inbox) readConv(ctx context.Context, uid gregor1.UID, convID chat1.ConversationID) (res types.RemoteConversation, err Error) {
+	convs, err := i.readConvs(ctx, uid, []chat1.ConversationID{convID})
 	if err != nil {
-		return NewInternalError(ctx, i.DebugLabeler, "failed to fetch server versions: %s", err.Error())
+		return res, err
 	}
-	ibox.ServerVersion = vers.InboxVers
-	ibox.Version = inboxVersion
-	i.summarizeConvs(ibox.Conversations)
-	i.Debug(ctx, "writeDiskInbox: uid: %s version: %d disk version: %d server version: %d convs: %d",
-		uid, ibox.InboxVersion, ibox.Version, ibox.ServerVersion, len(ibox.Conversations))
-	inboxMemCache.Put(uid, &ibox)
-	switch i.flushMode {
-	case InboxFlushModeActive:
-		return i.flushLocked(ctx, uid)
-	case InboxFlushModeDelegate:
-		return nil
+	return convs[0], nil
```

Comment:
can you sanity check the length before indexing? i know the other method returns a `MissError`, but to be future safe